### PR TITLE
Etcd encryption feature refactor for deployment and upgrades

### LINKED ISSUE #1399: [BUG] Epicli upgrade issue - the process hangs for several hours on the task kubeadm upgrade apply
**Describe the bug**
Cannot upgrade the kubernetes cluster - looks like a random issue, most common for Azure configurations.

**To Reproduce**
1. deploy 0.4.4 cluster
2. execute `epicli upgrade` (from develop branch)

**Expected behavior**
The cluster has been successfully upgraded

**OS (please complete the following information):**
 - OS: [e.g. RHEL7.7, Ubuntu 18.04]

**Cloud Environment (please complete the following information):**
 - Cloud Provider [e.g. MS Azure]

**Additional context**
```

2020-06-29T22:04:48.1026877Z 22:04:48 INFO cli.engine.ansible.AnsibleCommand - TASK [upgrade : upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file)] ***
2020-06-29T22:49:14.7940500Z 22:49:14 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (20 retries left).
2020-06-29T23:37:26.7290995Z 23:37:26 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (19 retries left).
2020-06-29T23:42:59.1434418Z 23:42:59 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (18 retries left).
2020-06-30T00:31:11.0432338Z 00:31:11 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (17 retries left).
2020-06-30T00:36:43.4952279Z 00:36:43 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (16 retries left).
2020-06-30T01:24:55.4254331Z 01:24:55 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (15 retries left).
2020-06-30T01:30:27.8367170Z 01:30:27 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (14 retries left).
2020-06-30T02:18:39.8409973Z 02:18:39 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (13 retries left).
2020-06-30T02:24:12.2573132Z 02:24:12 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (12 retries left).
2020-06-30T03:12:24.2749881Z 03:12:24 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (11 retries left).
2020-06-30T03:17:56.7150480Z 03:17:56 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (10 retries left).
2020-06-30T04:06:08.6468293Z 04:06:08 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (9 retries left).
2020-06-30T04:11:41.0625442Z 04:11:41 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (8 retries left).
2020-06-30T04:59:53.0070741Z 04:59:53 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (7 retries left).
2020-06-30T05:05:25.4451723Z 05:05:25 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (6 retries left).
2020-06-30T05:53:37.9468084Z 05:53:37 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (5 retries left).
2020-06-30T05:59:09.7340515Z 05:59:09 INFO cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: upgrade-master | Upgrade K8s cluster to v1.15.10 (using kubeadm-config.yml file) (4 retries left).

```
```
E0629 11:49:57.067369       1 reflector.go:125] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:49:58.070606       1 reflector.go:125] k8s.io/client-go/informers/factory.go:133: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:49:58.072162       1 reflector.go:125] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:49:59.075549       1 reflector.go:125] k8s.io/client-go/informers/factory.go:133: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:49:59.075968       1 reflector.go:125] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:50:00.089012       1 reflector.go:125] k8s.io/client-go/informers/factory.go:133: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:50:00.091010       1 reflector.go:125] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to list *v1.Secret: illegal base64 data at input byte 3
E0629 11:50:01.093364       1 reflector.go:125] k8s.io/client-go/informers/factory.go:133: Failed to list *v1.Secret: illegal base64 data at input byte 3

```

Looks like an issue with etcd encryption.

### PR DESCRIPTION
**Pull request by sk4zuzu**
Fixes issues with long running upgrades completely or partially (it still needs to be determined in a comprehensive testing :+1:).

### CODE REVIEW

**Comment by mkyc** [comment]
@sk4zuzu please change PR description not to invoke GH automation with "resolves #XXX" pattern

**Comment by sk4zuzu** [comment]
> @sk4zuzu please change PR description not to invoke GH automation with "resolves #XXX" pattern

@mkyc I changed the description and linked the issue manually :+1:.

**Comment by seriva** [comment]
Missing changelog entry?

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/patch-kubeadm-config.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      shell: |
+        kubectl get configmap kubeadm-config \
+          --namespace kube-system \
+          --output yaml
+      args:
+        executable: /bin/bash
+      environment:
+        KUBECONFIG: &KUBECONFIG /etc/kubernetes/admin.conf
+      register: shell_kubeadm_configmap
+      changed_when: false
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/patch-kubeadm-config.yml:89]: Please add logic to check whether state was changed or skip this task based on output of `kubeadm config view`

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/upgrade-kubeadm-config.yml [hunk 2]
```diff
@@ -1,7 +1,58 @@
...
-    line: "kubernetesVersion: v{{ version }}"
-    state: present
\ No newline at end of file
+- name: Check the kubeadm-config.yml file
+  stat:
+    path: &kubeadm-config-yml /etc/kubeadm/kubeadm-config.yml
+    get_attributes: false
+    get_checksum: false
+    get_mime: false
+  register: stat_kubeadm_config
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/upgrade-kubeadm-config.yml:17]:
```suggestion
  register: stat_kubeadm_config_file
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/upgrade-kubeadm-config.yml [hunk 3]
```diff
@@ -1,7 +1,58 @@
...
+
+- when: stat_kubeadm_config.stat.exists
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/upgrade-kubeadm-config.yml:19]:
```suggestion
- when: stat_kubeadm_config_file.stat.exists
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/patch-kubeadm-config.yml (hunk 1 above)
Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/patch-kubeadm-config.yml:89]: @to-bar please check this commit https://github.com/epiphany-platform/epiphany/pull/1427/commits/29bb389f9bd431f3ee239e435bd781708716fdfa I hope it solves the issue :+1:.

**Comment by sk4zuzu** [comment]
> Missing changelog entry?

@seriva it's not over yet. We fixed only one issue so far, it seems that we have at least one more that's causing similar problems. We will produce more PRs, please let us add the entry in the last one when we are sure it's finally fixed. :)

**Comment by mkyc** [comment]
> > Missing changelog entry?
> 
> @seriva it's not over yet. We fixed only one issue so far, it seems that we have at least one more that's causing similar problems. We will produce more PRs, please let us add the entry in the last one when we are sure it's finally fixed. :)

Why not add line: 

` - fixing Etcd encryption (#1427)`

and then add next PRs numbers? 

` - fixing Etcd encryption (#1427, #2222)`

and so on: 

` - fixing Etcd encryption (#1427, #2222, #3333)`

I think adding line to change log every PR is good practice.

**Comment by sk4zuzu** [comment]
> > > Missing changelog entry?
> > 
> > 
> > @seriva it's not over yet. We fixed only one issue so far, it seems that we have at least one more that's causing similar problems. We will produce more PRs, please let us add the entry in the last one when we are sure it's finally fixed. :)
> 
> Why not add line:
> 
> ` - fixing Etcd encryption (#1427)`
> 
> and then add next PRs numbers?
> 
> ` - fixing Etcd encryption (#1427, #2222)`
> 
> and so on:
> 
> ` - fixing Etcd encryption (#1427, #2222, #3333)`
> 
> I think adding line to change log every PR is good practice.

@to-bar, @rafzei and me discussed this briefly and as it maybe looks nice, we don't really see any benefit from adding it in each PR (extra work). As long as we propely link PRs to issues, there should be no problem in finding what was done at what point.

**Pull request merged**